### PR TITLE
Remove dependency to spree_auth_devise

### LINKED ISSUE #4014: Make OFN independent of spree_auth_devise
### What we should change and why (this is tech debt)
There's a mess in spree auth devise decorating spree code and then ofn decorating it on top and lots of code in spree_auth_devise we dont use in ofn.
There's great simplicity added if we simply remove spree_auth_devise and delete the code we dont use in ofn.

### Context
see #4013

### Impact and timeline
Very good opportunity to clean up user, registration, sessions controllers and views.

### PR DESCRIPTION
**Pull request by luisramos0**
#### What? Why?

Closes #4014

We are removing the spree authentication module and moving what we need to OFN codebase.
There are no additions and very few modifications, this is a careful copy paste exercise.

I thought of making multiple PRs for this but then realized this would not be too big and it's much better to review and test this all in one go.

I fixed a lot of rubocop issues in this PR. I leave a few unresolved.

#### What should we test?
Quite a few things to be carefully tested in this PR.
- [x] - admin/users: list, search, create, edit (including only reseting password), delete, generate_api_key, clear_api_key and regenerate_api_key (in this PR we fix a bug: if you try to delete an user with orders, instead of getting the snail, you get an error message that can be translated)
- [x] - checkout: guest checkout and login before checkout, checkout with already_registered email should not be possible
- [x] - user account: create account, login, forgot password, show list of orders and change password in the frontoffice account
- [x] - user sessions: logout frontoffice and backoffice
- [x] - user_mailer: reset password instructions email, signup confirmation email, account confirmation instructions (verify email sender address)
- [x] - backoffice top nav bar: link to shopfront and account
- [x] - edit customer details in an order
- [x] - detect missing translations in all these tests

#### Release notes
Changelog Category: Changed
Removed dependency to spree_auth_devise, the spree adapted version of devise, an authentication framework. OFN now works directly with devise.

### CODE REVIEW

File: app/controllers/spree/admin/base_controller_decorator.rb [hunk 1]
```diff
@@ -47,6 +47,16 @@ def unauthorized
...
+  protected
+
+  def model_class
+    const_name = controller_name.classify
+    if Spree.const_defined?(const_name)
+      return "Spree::#{const_name}".constantize
+    end
+    nil
+  end
+
```
Comment (luisramos0) [inline app/controllers/spree/admin/base_controller_decorator.rb]: this comes from spree_auth_devise admin_controller_decorator.rb

File: app/controllers/spree/admin/resource_controller_decorator.rb [hunk 2]
```diff
@@ -14,3 +14,7 @@ def load_resource
 end
 
 Spree::Admin::ResourceController.prepend(AuthorizeOnLoadResource)
+
+Spree::Admin::ResourceController.class_eval do
+  rescue_from CanCan::AccessDenied, :with => :unauthorized
+end
```
Comment (luisramos0) [inline app/controllers/spree/admin/resource_controller_decorator.rb]: this comes from spree_auth_devise admin_resource_controller_decorator.rb

File: app/controllers/spree/user_registrations_controller.rb [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+      super
+    end
+
+    protected
+
+    def check_permissions
+      authorize!(:create, resource)
+    end
+  end
+end
```
Comment (luisramos0) [inline app/controllers/spree/user_registrations_controller.rb]: I created #4019 to represent the task to delete some of all of this code from UserRegistrationsController and UserPasswordsController. For next time...

File: app/mailers/spree/user_mailer.rb [hunk 4]
```diff
@@ -8,5 +10,34 @@ def reset_password_instructions(user)
...
+
+    def signup_confirmation(user)
+      @user = user
+      I18n.with_locale valid_locale(@user) do
+        mail(to: user.email, from: from_address,
+             subject: t(:welcome_to) + Spree::Config[:site_name])
+      end
+    end
+
+    # Overriding `Spree::UserMailer.confirmation_instructions` which is
```
Comment (mkllnk) [inline app/mailers/spree/user_mailer.rb]: Hm, I expected this to replace something. Did spree_auth_devise stop overriding this? Does this comment need adjusting?

File: config/initializers/devise.rb [hunk 5]
```diff
@@ -1,5 +1,146 @@
...
+  # Set this to true to use Basic Auth for AJAX requests.  True by default.
+  #config.http_authenticatable_on_xhr = false
+
+  # The realm used in Http Basic Authentication
+  config.http_authentication_realm = 'Spree Application'
+
+  # ==> Configuration for :database_authenticatable
+  # For bcrypt, this is the cost for hashing the password and defaults to 10. If
+  # using other encryptors, it sets how many times you want the password re-encrypted.
+  config.stretches = 20
```
Comment (mkllnk) [inline config/initializers/devise.rb]: We should reduce that to a minimum in specs and see if that speeds them up.

File: Gemfile [hunk 6]
```diff
@@ -24,7 +24,6 @@ gem 'spree_api', github: 'openfoodfoundation/spree', branch: '2-0-4-stable'
 gem 'spree_backend', github: 'openfoodfoundation/spree', branch: '2-0-4-stable'
 gem 'spree_core', github: 'openfoodfoundation/spree', branch: '2-0-4-stable'
 
-gem 'spree_auth_devise', github: 'luisramos0/spree_auth_devise', branch: '2-0-without-ctrls'
```
[the remark this answers is not in the document]
Comment (luisramos0) [reply Gemfile]: yes, but this was my method to work this out: remove the dependency and work on the broken specs one by one.

File: app/mailers/spree/user_mailer.rb (hunk 4 above)
Comment (luisramos0) [reply app/mailers/spree/user_mailer.rb]: I'll have a close look at this.

File: config/initializers/devise.rb (hunk 5 above)
Comment (luisramos0) [reply config/initializers/devise.rb]: ok, I'll create a tech debt issue for that.

File: app/mailers/spree/user_mailer.rb (hunk 4 above)
Comment (luisramos0) [reply app/mailers/spree/user_mailer.rb]: this method is not there in v2.0, only introduced in v2.4 [here](https://github.com/openfoodfoundation/spree_auth_devise/commit/c25cfa1812438d41332f7e1545ea56fc1ce3c01d)
Anyway, I added a new commit to fix the comments, see [new commit](https://github.com/openfoodfoundation/openfoodnetwork/pull/3700/commits/942824cd74ca4528d2f58c507352bddc6a94d49c).

File: app/mailers/spree/user_mailer.rb [hunk 7]
```diff
@@ -12,6 +15,7 @@ def reset_password_instructions(user)
              I18n.t(:subject, scope: [:devise, :mailer, :reset_password_instructions]))
     end
 
+    # This is a OFN specific email, not from Devise::Mailer
```
Comment (mkllnk) [inline app/mailers/spree/user_mailer.rb]: Thank you!

**Comment by luisramos0** [comment]
I moved some specs from spree_auth_devise and found a bug in the process: #4031

File: config/initializers/devise.rb (hunk 5 above)
Comment (luisramos0) [reply config/initializers/devise.rb]: I went for it doing this: https://github.com/plataformatec/devise/wiki/Speed-up-your-unit-tests

**Comment by sauloperez** [comment]
Given

> I fixed a lot of rubocop issues in this PR. I leave a few unresolved.

Unless someones disagrees, I'll go ahead and approve the PR in CodeClimate.

File: config/initializers/devise.rb [hunk 8]
```diff
@@ -37,7 +37,7 @@
   # ==> Configuration for :database_authenticatable
   # For bcrypt, this is the cost for hashing the password and defaults to 10. If
   # using other encryptors, it sets how many times you want the password re-encrypted.
-  config.stretches = 20
+  config.stretches = Rails.env.test? ? 1 : 20
```
Comment (mkllnk) [inline config/initializers/devise.rb]: And? Did it help? Semaphore actually ran 2.5 minutes longer. But one run each is not a reliable measure.

Comment (luisramos0) [reply config/initializers/devise.rb]: the build times are too unstable to see, the 5 builds before this change had times between 51 and 1h04 minutes...

**Comment by luisramos0** [comment]
see https://github.com/openfoodfoundation/ofn-install/issues/476
I have added a commit to bring db/default/user.rb, with this we would have trouble in ofn-install provisioning servers for the frist time.

**Comment by lin-d-hop** [comment]
Testing notes here:
https://docs.google.com/document/d/1PlzyBYSIVXoDgd-ZBwqS7sRQFZzM6QwrDcNeeLEizeM/edit#

**Comment by lin-d-hop** [comment]
@luisramos0 @rachL I've been going through the tests and just noticed the last one is 'Detact missing translations'. As I've been doing all the testing in English I've definitely not been noticing missing translations. 
Would you both be happy for this to go through without that check or do we need to retest? :'-(

**Comment by lin-d-hop** [comment]
Aight, gone through an tested it all.... all looks great!
Just awaiting on a response regarding the criticality of missing translations. If we don't mind a potential missed translation then it is Ready to Go!

**Comment by luisramos0** [comment]
thanks @lin-d-hop for all the testing! I am very happy this PR passed the tests, I think this is one of my favorite PRs ever 😍

**Comment by RachL** [comment]
woooow you've tested eveyrthing, you rock!!!!

Usually "missing translations" are missing keys, so you should have seen them in English as well :) So I'm moving this to the ready to go column, awesome!

**Pull request merged**